### PR TITLE
Add RawSortedIndexBasedFilterOperator for binary search on raw sorted columns

### PR DESCRIPTION
**Pull request by xiangfu0**
## Summary
- Add `RawSortedIndexBasedFilterOperator` that uses binary search O(log N) on raw sorted forward index columns instead of full scan O(N)
- For chunk-compressed readers, implements two-level binary search: coarse search at chunk boundaries minimizes decompressions, fine search within cached chunk is free
- Add `MultiChunkReaderContext` with an LRU cache of decompressed chunks to avoid repeated decompression during binary search; chunk cache is exposed via `ForwardIndexReader.createCachedContext(int numSlots)` SPI
- Supports EQ, NEQ, IN, NOT_IN, RANGE predicates with all data types (INT, LONG, FLOAT, DOUBLE, STRING, BYTES, BIG_DECIMAL)
- Exposes `getNumDocsPerChunk()` and `createCachedContext()` on `ForwardIndexReader` interface for chunk-aware optimization

## Background
Currently `SortedIndexBasedFilterOperator` only works for dictionary-encoded sorted columns. Raw sorted columns fall through to `ScanBasedFilterOperator` which linearly scans every document. This PR adds an efficient binary search path for raw sorted columns, matching the optimization that dictionary-encoded sorted columns already enjoy.

## Changes
- `ForwardIndexReader.java` — Added `getNumDocsPerChunk()` and `createCachedContext(int numSlots)` default methods
- `BaseChunkForwardIndexReader.java` — Override both methods; adds `decompressChunkInto` for direct-into-slot decompression
- `MultiChunkReaderContext.java` — New LRU chunk cache context (up to N decompressed chunks); handles DELTA/DELTADELTA codecs correctly
- `RawSortedIndexBasedFilterOperator.java` — New filter operator with two-level binary search + multi-chunk cache; caches `computeMatchingRanges()` result for reuse by `canOptimizeCount()` / `canProduceBitmaps()`
- `FilterOperatorUtils.java` — Route sorted raw SV columns to the new operator
- `RawSortedIndexBasedFilterOperatorTest.java` — 21 unit tests covering all predicate types and edge cases
- `MultiChunkReaderContextTest.java` — 8 unit tests covering cache hit/miss, LRU eviction, replaceSlot, close cleanup, and integration for all compression types
- `BenchmarkRawSortedIndexFilter.java` — JMH microbenchmark comparing binary search vs linear scan

## Benchmark

JMH microbenchmark on a sorted raw INT forward index (1K docs/chunk, ~10x value repetition).
- EQ: matches median value; RANGE: ~1% selectivity window around median
- Machine: Apple M-series, JDK 17, 1 fork, 3 warmup + 5 measurement × 1s, `AverageTime`

```
Benchmark                                       compression    numDocs    Score    Error  Units
-----------------------------------------------------------------------------------------------
binarySearchEq                                 PASS_THROUGH  1,000,000    5.6 ±  2.2  us/op
binarySearchEq                                 PASS_THROUGH  5,000,000    5.7 ±  3.3  us/op
binarySearchEq                                 PASS_THROUGH 10,000,000    5.5 ±  3.2  us/op
linearScanEq   (baseline)                      PASS_THROUGH  1,000,000   22.7 ±  5.3  us/op
linearScanEq   (baseline)                      PASS_THROUGH  5,000,000   22.8 ±  7.1  us/op
linearScanEq   (baseline)                      PASS_THROUGH 10,000,000   24.3 ± 10.7  us/op

binarySearchEq                                          LZ4  1,000,000   13.4 ±  1.1  us/op
binarySearchEq                                          LZ4  5,000,000   14.7 ±  1.4  us/op
binarySearchEq                                          LZ4 10,000,000   15.3 ±  0.8  us/op
linearScanEq   (baseline)                               LZ4  1,000,000   23.7 ±  4.9  us/op
linearScanEq   (baseline)                               LZ4  5,000,000   24.3 ±  6.3  us/op
linearScanEq   (baseline)                               LZ4 10,000,000   25.0 ± 11.3  us/op

binarySearchEq   (with chunk cache)               ZSTANDARD  1,000,000   27.4 ±  0.9  us/op
binarySearchEq   (with chunk cache)               ZSTANDARD  5,000,000   31.7 ±  1.8  us/op
binarySearchEq   (with chunk cache)               ZSTANDARD 10,000,000   34.2 ±  2.2  us/op
linearScanEq   (baseline)                         ZSTANDARD  1,000,000   22.7 ±  2.9  us/op
linearScanEq   (baseline)                         ZSTANDARD  5,000,000   25.7 ± 20.0  us/op
linearScanEq   (baseline)                         ZSTANDARD 10,000,000   24.5 ±  9.6  us/op

binarySearchRange                              PASS_THROUGH  1,000,000    5.6 ±  2.7  us/op
binarySearchRange                              PASS_THROUGH  5,000,000    5.6 ±  3.7  us/op
binarySearchRange                              PASS_THROUGH 10,000,000    5.7 ±  2.2  us/op
linearScanRange  (baseline)                    PASS_THROUGH  1,000,000   23.4 ±  5.2  us/op
linearScanRange  (baseline)                    PASS_THROUGH  5,000,000   24.2 ±  6.5  us/op
linearScanRange  (baseline)                    PASS_THROUGH 10,000,000   23.5 ±  5.9  us/op

binarySearchRange                                       LZ4  1,000,000   20.7 ±  0.6  us/op
binarySearchRange                                       LZ4  5,000,000   23.2 ±  0.6  us/op
binarySearchRange                                       LZ4 10,000,000   25.8 ±  1.6  us/op
linearScanRange  (baseline)                             LZ4  1,000,000   23.3 ±  5.2  us/op
linearScanRange  (baseline)                             LZ4  5,000,000   22.9 ±  3.5  us/op
linearScanRange  (baseline)                             LZ4 10,000,000   25.5 ± 14.4  us/op

binarySearchRange  (with chunk cache)             ZSTANDARD  1,000,000   47.8 ±  1.3  us/op  ⚠️
binarySearchRange  (with chunk cache)             ZSTANDARD  5,000,000   56.4 ±  1.5  us/op  ⚠️
binarySearchRange  (with chunk cache)             ZSTANDARD 10,000,000   62.4 ±  1.7  us/op  ⚠️
linearScanRange  (baseline)                       ZSTANDARD  1,000,000   23.9 ±  7.8  us/op
linearScanRange  (baseline)                       ZSTANDARD  5,000,000   25.2 ± 16.5  us/op
linearScanRange  (baseline)                       ZSTANDARD 10,000,000   23.7 ±  6.7  us/op
```

**Key takeaways:**
- **Uncompressed (PASS_THROUGH):** binary search is **~4× faster** (~5–6 µs vs ~23 µs). Cost is flat across 1M–10M docs.
- **LZ4:** EQ is **~1.7× faster** (~13–15 µs vs ~24 µs); RANGE is **comparable** (~21–26 µs vs ~23 µs) — the two binary searches (lower + upper bound) probe mostly disjoint chunk sets, limiting cache reuse.
- **ZSTANDARD/EQ:** chunk cache reduces cost from ~42 µs (without cache) to **~27–34 µs** (vs ~24 µs linear), a 35% improvement. EQ's lower+upper bound searches overlap on chunks near the match point, giving the cache meaningful hits.
- **ZSTANDARD/RANGE:** binary search (~48–62 µs) is slower than linear scan (~24 µs). RANGE's lower and upper bound searches probe opposite ends of the index with almost no chunk overlap, so the cache cannot help. ZSTANDARD's per-decompression cost × ~26 distinct chunks searched exceeds a linear scan's first batch.
- Linear scan time is nearly independent of `numDocs` because `ScanBasedFilterOperator` uses batched iteration — only the first batch (~1K docs) is measured here. Real query latency at scale is proportionally worse for linear scan.

## Test plan
- [x] 21 new unit tests (EQ, NEQ, IN, NOT_IN, RANGE with inclusive/exclusive/unbounded, chunk-aware search, all data types, edge cases)
- [x] 8 new unit tests for `MultiChunkReaderContext` (cache hit/miss, LRU eviction, replaceSlot, close, all compression types)
- [x] 96 existing FilterOperatorUtils tests pass
- [x] Checkstyle clean
- [x] Spotless formatted

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/operator/filter/FilterOperatorUtils.java [hunk 1]
```diff
@@ -104,6 +102,11 @@ public BaseFilterOperator getLeafFilterOperator(QueryContext queryContext, Predi
             && queryContext.isIndexUseAllowed(dataSource, FieldConfig.IndexType.RANGE)) {
           return new RangeIndexBasedFilterOperator(queryContext, predicateEvaluator, dataSource, numDocs);
         }
+        if (dataSource.getDataSourceMetadata().isSorted() && dataSource.getDictionary() == null
+            && dataSource.getDataSourceMetadata().isSingleValue()
+            && queryContext.isIndexUseAllowed(dataSource, FieldConfig.IndexType.SORTED)) {
+          return new RawSortedIndexBasedFilterOperator(queryContext, predicateEvaluator, dataSource, numDocs);
+        }
```
Comment (lakechd) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/filter/FilterOperatorUtils.java:109]: this is duplicated with line 140
